docs: Add more detail on how to use grafana with n8n metrics (#4628)
n8n-io/n8n-docs: docs/hosting/grafana/grafana.md

---

> contentType: howto

Author: These docs only become relevant if and once https://github.com/n8n-io/n8n/pull/30416 is merged

---

# Grafana

Enable the Prometheus metrics endpoint in n8n, then connect Grafana through a Prometheus instance to visualize your n8n data. Refer to [Enable Prometheus metrics](../configuration/configuration-examples/prometheus.md) before following this guide.

/// info | Feature availability
The `/metrics` endpoint isn't available on n8n Cloud.
///

## Configure Prometheus to scrape n8n

Prometheus scrapes n8n's `/metrics` endpoint on a schedule and stores the data. Grafana then queries Prometheus to display it.

> Add a scrape job for n8n in your `prometheus.yml`:

Reviewer: ```suggestion
Add a scrape job targeting your n8n instance in your `prometheus.yml`:
```

```yaml
scrape_configs:
	- job_name: n8n
		static_configs:
			- targets:
					- <n8n-host>:<n8n-port>
		metrics_path: /metrics
```

Replace `<n8n-host>` and `<n8n-port>` with your n8n instance address. The default n8n port is `5678`.

After editing, reload or restart Prometheus to apply the change.

## Add Prometheus as a Grafana data source

1. In Grafana, go to **Connections** > **Data sources**.
2. Select **Add new data source**.
3. Select **Prometheus**.
4. Set **Prometheus server URL** to your Prometheus instance address, for example `http://prometheus:9090`.
5. Select **Save & test**.

Grafana confirms the connection with a success message.

## Webhook observability

> n8n exposes a `n8n_webhook_request_duration_seconds` histogram for every webhook call. Enable these environment variables to collect it:

Reviewer: A question I asked when reading this is why is this a configurable property? Does it have an impact on performance, memory usage or some other part of the system when enabled? 

edit: Reading further down I think it's security and "label cardinality" because we expose the webhook path?

Author: Prometheus metrics are "opt-in by default". Each metric we implement in code has to manually enabled by our customers. See https://github.com/n8n-io/n8n/blob/master/packages/%40n8n/config/src/configs/endpoints.config.ts
Only `CPU, memory, GC, etc.` metrics are enabled out of the box.

Author: For adoption: I think it'd be nice to have a settings page for configuring prometheus metrics too (as we do for otel now).
The thing is: that UI would also have to be enabled manually. Because the `/metrics` endpoint is not reachable on our cloud instances.
Cloud customers would need to setup some auth token first in order for this to be exposed safely.

Author: In the meantime, I'm working to move https://github.com/konstantintieber/n8n-grafana-templates into the n8n org.
Each "shared dashboard" would list the required metrics that need to be enabled for it to work.

| p95 latency across all webhooks | `histogram_quantile(0.95, sum by (le) (rate(n8n_webhook_request_duration_seconds_bucket[5m])))` |
| p95 latency per workflow | `histogram_quantile(0.95, sum by (le, workflow_id) (rate(n8n_webhook_request_duration_seconds_bucket[5m])))` |
| Error rate (non-2xx) | `sum by (workflow_id) (rate(n8n_webhook_request_duration_seconds_count{status_code!="200"}[5m]))` |
| Average request duration | `rate(n8n_webhook_request_duration_seconds_sum[5m]) / rate(n8n_webhook_request_duration_seconds_count[5m])` |

## Form submission observability

n8n exposes a `n8n_form_submission_duration_seconds` histogram for every form submission. Enable these environment variables to collect it:

| Variable | Default | Description |
|----------|---------|-------------|
| `N8N_METRICS_INCLUDE_FORM_METRICS` | `false` | Exposes the `n8n_form_submission_duration_seconds` histogram. |
| `N8N_METRICS_INCLUDE_WORKFLOW_INFO` | `false` | Exposes the `n8n_workflow_info` gauge for human-readable workflow names. See [Workflow name lookup](#workflow-name-lookup). |

### What the metric tracks

> `n8n_form_submission_duration_seconds` is a Prometheus histogram. For each form submission, n8n records how long processing took and increments the matching latency buckets. The metric exposes three series per label combination:

Author: Note to self: change this to "how long it took to give the user feedback that the form was submitted".
We don't want to create the impression that this duration here equals workflow execution.

---
contentType: howto
---

# Grafana

Enable the Prometheus metrics endpoint in n8n, then connect Grafana through a Prometheus instance to visualize your n8n data. Refer to [Enable Prometheus metrics](/hosting/configuration/configuration-examples/prometheus.md) before following this guide.

/// info | Feature availability
The `/metrics` endpoint isn't available on n8n Cloud.
///

## Reusable dashboard templates

Once you enable Prometheus metrics for your n8n instance, you'll want to build dashboards to observe them.

n8n publishes reusable grafana dashboards for several of the supported prometheus metrics in this GitHub project:
[n8n-observability](https://github.com/n8n-io/n8n-observability){:target="_blank" .external-link}

## Configure Prometheus to scrape n8n

Prometheus scrapes n8n's `/metrics` endpoint on a schedule and stores the data. Grafana then queries Prometheus to display it.

Add a scrape job targeting your n8n instance in your `prometheus.yml`:

```yaml
scrape_configs:
  - job_name: n8n
    static_configs:
      - targets:
          - <n8n-host>:<n8n-port>

>     metrics_path: /metrics

Author: Fixed in https://github.com/n8n-io/n8n-docs/pull/4628/commits/86049f6059095e1a529e766791c9dda6c3c159ea

---
contentType: howto
---

# Grafana

> Enable the Prometheus metrics endpoint in n8n, then connect Grafana through a Prometheus instance to visualize your n8n data. Refer to [Enable Prometheus metrics](../configuration/configuration-examples/prometheus.md) before following this guide.

Reviewer 2: ```suggestion
Enable the Prometheus metrics endpoint in n8n, then connect Grafana through a Prometheus instance to visualize your n8n data. Refer to [Enable Prometheus metrics](/hosting/configuration/configuration-examples/prometheus.md) before following this guide.
```

/// info | Feature availability
The `/metrics` endpoint isn't available on n8n Cloud.
///

## Reusable dashboard templates

> Once you enabled prometheus metrics for your n8n instance, you'll want to build dashboards to observe them.

Reviewer 2: ```suggestion
Once you enable Prometheus metrics for your n8n instance, you'll want to build dashboards to observe them.
```

n8n publishes reusable grafana dashboards for several of the supported prometheus metrics in this GitHub project:

> [n8n-observability](https://github.com/n8n-io/n8n-observability)

Reviewer 2: ```suggestion
[n8n-observability](https://github.com/n8n-io/n8n-observability){:target="_blank" .external-link}
```

## Configure Prometheus to scrape n8n

Prometheus scrapes n8n's `/metrics` endpoint on a schedule and stores the data. Grafana then queries Prometheus to display it.

Add a scrape job targeting your n8n instance in your `prometheus.yml`:

> ```yaml
> scrape_configs:
> 	- job_name: n8n
> 		static_configs:
> 			- targets:
> 					- <n8n-host>:<n8n-port>
> 		metrics_path: /metrics
> ```

Reviewer 2: ```suggestion
```yaml
scrape_configs:
  - job_name: n8n
    static_configs:
      - targets:
          - <n8n-host>:<n8n-port>
    metrics_path: /metrics
```
```

Replace `<n8n-host>` and `<n8n-port>` with your n8n instance address. The default n8n port is `5678`.

After editing, reload or restart Prometheus to apply the change.

## Add Prometheus as a Grafana data source

1. In Grafana, go to **Connections** > **Data sources**.
2. Select **Add new data source**.
3. Select **Prometheus**.
4. Set **Prometheus server URL** to your Prometheus instance address, for example `http://prometheus:9090`.
5. Select **Save & test**.

Grafana confirms the connection with a success message.

## Webhook observability

Available from n8n version 2.28.0.

n8n exposes a `n8n_webhook_request_duration_seconds` histogram for every webhook call. Enable these environment variables to collect it:

| Variable | Default | Description |
|----------|---------|-------------|
| `N8N_METRICS_INCLUDE_WEBHOOK_METRICS` | `false` | Exposes the `n8n_webhook_request_duration_seconds` histogram. |
| `N8N_METRICS_INCLUDE_WORKFLOW_INFO` | `false` | Exposes the `n8n_workflow_info` gauge for human-readable workflow names. See [Workflow name lookup](#workflow-name-lookup). |

> You can find a [ready-to-use dashboard](https://github.com/n8n-io/n8n-observability/tree/main/dashboards/grafana/n8n-webhook-executions) for this histogram in our n8n-observability project.

Reviewer 2: ```suggestion
You can find a [ready-to-use dashboard](https://github.com/n8n-io/n8n-observability/tree/main/dashboards/grafana/n8n-webhook-executions){:target="_blank" .external-link} for this histogram in our n8n-observability project.
```

| p95 latency across all webhooks | `histogram_quantile(0.95, sum by (le) (rate(n8n_webhook_request_duration_seconds_bucket[5m])))` |
| p95 latency per workflow | `histogram_quantile(0.95, sum by (le, workflow_id) (rate(n8n_webhook_request_duration_seconds_bucket[5m])))` |
| Error rate (non-2xx) | `sum by (workflow_id) (rate(n8n_webhook_request_duration_seconds_count{status_code!="2.."}[5m]))` |
| Average request duration | `rate(n8n_webhook_request_duration_seconds_sum[5m]) / rate(n8n_webhook_request_duration_seconds_count[5m])` |

## Form submission observability

Available from n8n version 2.28.0.

n8n exposes a `n8n_form_submission_duration_seconds` histogram for every form submission. Enable these environment variables to collect it:

| Variable | Default | Description |
|----------|---------|-------------|
| `N8N_METRICS_INCLUDE_FORM_METRICS` | `false` | Exposes the `n8n_form_submission_duration_seconds` histogram. |
| `N8N_METRICS_INCLUDE_WORKFLOW_INFO` | `false` | Exposes the `n8n_workflow_info` gauge for human-readable workflow names. See [Workflow name lookup](#workflow-name-lookup). |

> You can find a [ready-to-use dashboard](https://github.com/n8n-io/n8n-observability/tree/main/dashboards/grafana/n8n-form-executions) for this histogram in our n8n-observability project.

Reviewer 2: ```suggestion
You can find a [ready-to-use dashboard](https://github.com/n8n-io/n8n-observability/tree/main/dashboards/grafana/n8n-form-executions){:target="_blank" .external-link} for this histogram in our n8n-observability project.
```
